### PR TITLE
UW-559 Add argument for directing drivers to an arbitrary block of an experiment YAML.

### PR DESCRIPTION
**Pull request by NaureenBharwaniNOAA**
**Synopsis**

In a config we currently support a high-level key with the drivers name. 
i.e. 
```
mpas:
    ....
    ....
```

- We are adding an argument to support the driver being nested further within the config by traversing down a set of keys.
- We also want to support a driver running multiple times and may have a different naming scheme than one chosen.
i.e.
```
foo:
    mpas:
        ....
        ....
bar:
    mpas:
        ....
        ....
``` 

**Type**

- [x] Enhancement (adds new functionality)

**Impact**

- [x] This is a non-breaking change (existing functionality continues to work as expected)

**Checklist**

- [x] I have added myself and any co-authors to the PR's _Assignees_ list.
- [x] I have reviewed the documentation and have made any updates necessitated by this change.

### CODE REVIEW

File: src/uwtools/drivers/make_solo_mosaic.py
Comment (NaureenBharwaniNOAA) [inline src/uwtools/drivers/make_solo_mosaic.py:34]: Add key_path argument here

File: src/uwtools/tests/test_cli.py
Comment (NaureenBharwaniNOAA) [inline src/uwtools/tests/test_cli.py:565]: Trying to figure out how to come up with a test case for this currently.

**Comment by maddenp-noaa** [comment]
A somewhat pedantic point maybe, but is this really a breaking change? Wouldn't everything that already worked continue to work? This seems like only additional functionality that users could just ignore if they didn't need it.

**Comment by maddenp-noaa** [comment]
I don't see any doc updates, and the _I have reviewed the documentation and have made any updates necessitated by this change._ box is not ticked. I think we'd at least need to update the CLI `--help` examples to reflect the new option*, and probably we'd want a simple use example in the CLI docs.

* Feel free to ignore updating the `--help` output: I'll have a PR up soon that automates the generation of that content so that it will be easier for all of us going forward. It would be a waste of your time to manually update it right now. But I think a use example in the docs would still be helpful. Or do we have a different ticket for that work?

File: src/uwtools/cli.py
Comment (maddenp-noaa) [reply src/uwtools/cli.py:887]: This is where simple built-in integration tests would come in handy: If you'd added an integration test that executed, say, `uw upp --config-file config.yaml --key-path a.b [other options]`, it would fail because, without this line, `--key-path` would not be recognized.

But then I have to ask: How did you successfully test this new feature from the CLI without this line? I don't think you could have. What works best for me is doing the code development and manually testing the feature -- which for a CLI-exposed feature like this one would necessitate a command-line test -- until it appears to behave as I expect, and _then_ write unit tests, and _then_ write integration tests (someday soon!), and _then_ write documentation.

File: src/uwtools/drivers/chgres_cube.py
Comment (maddenp-noaa) [inline src/uwtools/drivers/chgres_cube.py:37]: I wonder if this might benefit from a little more specificity: It's not just _any_ subsection of the config we're looking for (and it's not necessarily a file -- it may be data streamed on `stdin` -- so we should probably not use that word here), it's the driver-configuration block. Maybe something like
```
:param key_path: Keys leading through the config to the driver's configuration block.
```
? Here and in the other driver modules?

File: src/uwtools/drivers/driver.py
Comment (maddenp-noaa) [inline src/uwtools/drivers/driver.py:32]: I think we made this change to support the unit tests. Am I recalling correctly? I think it's a good change, and maybe some day we'll have reason to make the same change in the subclasses, but not today.

**Comment by NaureenBharwaniNOAA** [comment]
> I don't see any doc updates, and the _I have reviewed the documentation and have made any updates necessitated by this change._ box is not ticked. I think we'd at least need to update the CLI `--help` examples to reflect the new option*, and probably we'd want a simple use example in the CLI docs.
> 
> * Feel free to ignore updating the `--help` output: I'll have a PR up soon that automates the generation of that content so that it will be easier for all of us going forward. It would be a waste of your time to manually update it right now. But I think a use example in the docs would still be helpful. Or do we have a different ticket for that work?
> 
> **EDIT** Sorry, I realize now that this is for driver support, and it would be annoying to repeat an example of using `--key-path` for every driver. Maybe a generic use case could be constructed in a file in `docs/shared` and then that snipped included in each driver CLI page. It might show something like changing from
> 
> ```
> driver:
>     ... driver config ...
> ```
> 
> to
> 
> ```
> a:
>   b:
>     driver:
>       ... driver config ...
> ```
> 
> and how `--key-path a.b` could be used to select the correct driver.

I'll add this in the documentation. Originally when I made the draft PR I didn't add any yet as I was still figuring out the bool value to a List of strings. I'll update this soon.

File: src/uwtools/drivers/driver.py
Comment (NaureenBharwaniNOAA) [reply src/uwtools/drivers/driver.py:32]: Yes you are! 

It was this error message:
```
=> Running typechecker
+ mypy --install-types --non-interactive .
uwtools/tests/drivers/test_driver.py:123: error: Argument "config" to
"ConcreteDriver" has incompatible type "dict[str, dict[str, Any]]";
expected "Path | None"  [arg-type]
            config={"foo": {"bar": config}},
                   ^~~~~~~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 127 source files)
```

File: docs/sections/user_guide/cli/drivers/chgres_cube.rst
Comment (NaureenBharwaniNOAA) [inline docs/sections/user_guide/cli/drivers/chgres_cube.rst:100]: I added the `key_path` arg documentation here for a single driver. Once I get some feedback on how this looks, I can add it to all drivers. I also can factor all the information out to it's own section (not just the shared yaml) since this will be the same across all drivers.

File: docs/sections/user_guide/cli/drivers/chgres_cube.rst
Comment (NaureenBharwaniNOAA) [inline docs/sections/user_guide/cli/drivers/chgres_cube.rst:1]: @maddenp-noaa I didn't add the `--key-path` argument to the help output from `uw chgres_cube run --help` command here based on this comment:

> Feel free to ignore updating the --help output: I'll have a PR up soon that automates the generation of that content so that it will be easier for all of us going forward. It would be a waste of your time to manually update it right now. But I think a use example in the docs would still be helpful. Or do we have a different ticket for that work?

I believe that is #495 that you're currently working on. I see here that you factored it out into this [file](https://github.com/ufs-community/uwtools/pull/495/files#diff-0cdc92eda7b1121e63f56f7bedd12951e9b69114cf8274f607046755aa7046e3R1). I think the updating would depend on which merged first.

Comment (maddenp-noaa) [reply docs/sections/user_guide/cli/drivers/chgres_cube.rst:1]: I think it'll be easiest is you don't update the `--help` output, as it will create merge conflicts if you do. If I merge before you do, after you merge my changes onto your branch you'll just need to run `make docs` from your clone root and any changes related to your work in this PR should be reflected in the docs. Let's talk if my PR merges before yours.

File: docs/sections/user_guide/cli/drivers/chgres_cube.rst
Comment (maddenp-noaa) [reply docs/sections/user_guide/cli/drivers/chgres_cube.rst:100]: I think it would keep things DRYer if the explanatory text as well as the YAML example were include-able. For example, something like `key_path_nested.yaml`
```
foo:
  bar:
    driver:
      # driver config block
```
and `key_path_simple.yaml`
```
driver:
  # driver config block
```
and `key_path.rst`
```
The ``--key-path`` option can be used to navigate from the top of the config to the driver's configuration block. For example, specifying ``--key-path foo.bar`` with config

.. literalinclude:: shared/key_path_nested.yaml
.. highlight:: yaml

is equivalent to using config

.. literalinclude:: shared/key_path_simple.yaml
.. highlight:: yaml

without specifying ``--key-path``.
```
Then, I think, we could just add the single line
```
.. literalinclude:: ../../../../shared/key_path.rst
```
to each driver doc to get the explanatory text and nicely formatted YAML snippets. I'm not sure we have other examples of one include including another, so hopefully that's more-or-less right.

Does that proposed explanatory text make sense?

File: docs/shared/key_path.rst
Comment (NaureenBharwaniNOAA) [inline docs/shared/key_path.rst:4]: Failing the CI/CD run with an error: 
```
Warning, treated as error:
/Users/naureenbharwani/Desktop/UFS/uwtools/docs/shared/key_path.rst:6:Include file '/Users/naureenbharwani/Desktop/shared/key_path_nested.yaml' not found or reading it failed
```

The documentation is being generated correctly [here](https://uwtools--490.org.readthedocs.build/en/490/sections/user_guide/cli/drivers/chgres_cube.html).

File: docs/shared/key_path.rst
Comment (maddenp-noaa) [inline docs/shared/key_path.rst:1]:
```suggestion
  The ``--key-path`` option can be used to navigate from the top of the config to the driver's configuration block. For example, specifying ``--key-path foo.bar`` with config
```

File: docs/shared/key_path.rst
Comment (maddenp-noaa) [inline docs/shared/key_path.rst:6]:
```suggestion
  is equivalent to using config
```

File: docs/shared/key_path.rst
Comment (maddenp-noaa) [inline docs/shared/key_path.rst:11]:
```suggestion
  without specifying ``--key-path``.
```

File: docs/sections/user_guide/cli/drivers/chgres_cube.rst
Comment (maddenp-noaa) [inline docs/sections/user_guide/cli/drivers/chgres_cube.rst:97]: This looks good to me in the rendered docs. If no one else disagrees, I think we could add this line to the other driver `.rst` files, since they all inherit the `--key-path` functionality.

File: src/uwtools/cli.py
Comment (NaureenBharwaniNOAA) [inline src/uwtools/cli.py:644]: I wanted to point out that I changed the help message for the argument as well.

File: docs/shared/key_path.rst
Comment (maddenp-noaa) [reply docs/shared/key_path.rst:4]: I checked out your branch and had a look at this and think I know the solution. Try specifying
```
.. literalinclude:: /shared/key_path_nested.yaml
```
and
```
.. literalinclude:: /shared/key_path_simple.yaml
```
Apparently, and this was news to me, Sphinx interprets absolute paths as relative to "the top source directory" which, in our case, for docs, is `docs/`. So, for me, the above works.

Also, I think I misled you about the indentation: I think the lines in this file should be all the way to the left, but the line
```
.. include:: ../../../../shared/key_path.rst
```
in `chgres_cube.rst` should be indented to match its context.

**Review by maddenp-noaa — APPROVED** [review]
It all looks great, nice work.

File: src/uwtools/tests/test_cli.py
Comment (christinaholtNOAA) [inline src/uwtools/tests/test_cli.py:549]: This is an interesting use case I had not considered before. Is the logic implemented to turn this into the contents of the config.yaml file? I see that it doesn't quite matter for the test, but it might give off the wrong idea as an example, if the logic does not exist. 

I don't think this should hold up this PR, to be clear.

**Review by christinaholtNOAA — APPROVED** [review]
LGTM. Just one clarification question, but I'm happy with it as-is.

File: src/uwtools/tests/test_cli.py
Comment (NaureenBharwaniNOAA) [reply src/uwtools/tests/test_cli.py:549]: Great point, keeping the original text as is passes the tests as well. I originally added this in when debugging how to get 100% test coverage. I don't think it needs to be here and happy to remove if agree

Comment (maddenp-noaa) [reply src/uwtools/tests/test_cli.py:549]: It's slightly complicated: To support the tests in `uwtools.tests.drivers.test_driver` via the `ConcreteDriver(Driver)` class defined there, we updated the signature of the `Driver` base class constructor to accept a `dict`, to avoid the need to write to disk. I suspect this is where @NaureenBharwaniNOAA got the idea to use a `dict` in the current test. This is an obviously valid change, because `Driver` immediately passes this value to the `YAMLConfig` constructor, which is built to receive a `dict` (among other value types).

But all the `Driver` subclasses currently require `config` to be `Optional[Path]` and don't support a `dict` value. We could easily change that later if we found a need, but I don't see it now.

That said, in the current test, the driver object is a `Mock` anyway, so it doesn't matter what the value is. As @christinaholtNOAA suggests, using a `dict` here might be confusing. I'd suggest changing it to a `str` like `"/path/to/config"`. I could change it in a near-future omnibus PR if the change doesn't make it into this PR.

Comment (NaureenBharwaniNOAA) [reply src/uwtools/tests/test_cli.py:549]: Updating with this change now. I applied the change to the `test__dispatch_to_driver()` test case specifically. I believe that was the action to take.

**Pull request merged**
